### PR TITLE
Enable headers-only file to get headers using 'headers: true' option

### PR DESCRIPTION
**Pull request by shibasaka**
When trying to get headers from headers-only file using `headers: true` option, Table#headers method currently returns [].
So, I thought it's proper that Table#headers method return file's header in this case.

For example:
test.csv(headers-only file)
```
a,b,c
```

the previous behaviour
```
csv = CSV.read 'test.csv', {headers: true}
csv.headers
=> []
```

changed it as follows
```
csv = CSV.read 'test.csv', {headers: true}
csv.headers
=> ["a", "b", "c"]
```

I added an argument of headers to `Table.new` method because Table class should always has headers.

### CODE REVIEW

File: lib/csv/table.rb [hunk 1]
```diff
@@ -258,6 +259,11 @@ def delete(*indexes_or_headers)
             (@mode == :col_or_row and index_or_header.is_a? Integer)
           @table.delete_at(index_or_header)
         else                 # by header
+          if index_or_header.is_a? Integer
+            @headers.delete_at(index_or_header)
+          else
+            @headers.delete(index_or_header)
+          end
```
Comment (kou) [inline lib/csv/table.rb]: Can you add a test for this case?

File: lib/csv/table.rb [hunk 2]
```diff
@@ -285,7 +291,8 @@ def delete_if(&block)
         @table.delete_if(&block)
       else                                      # by header
         deleted = []
-        headers.each do |header|
+        initial_headers = headers.dup
```
Comment (kou) [inline lib/csv/table.rb]: Is `dup` needed?

Comment (shibasaka) [reply lib/csv/table.rb]: I've fixed it(because `dup` had been duplicated in `headers` method).

File: lib/csv/table.rb (hunk 1 above)
Comment (shibasaka) [reply lib/csv/table.rb]: I've added tests of deleting headers by index and header.

File: lib/csv/table.rb [hunk 3]
```diff
@@ -23,7 +23,8 @@ class Table
     # * length()
     # * size()
     #
-    def initialize(array_of_rows)
+    def initialize(array_of_rows, headers=nil)
+      @headers = headers || []
```
[the remark this answers is not in the document]
Comment (shibasaka) [reply lib/csv/table.rb]: I've changed it and restored argument in test code.

File: lib/csv/table.rb [hunk 4]
```diff
@@ -23,8 +23,19 @@ class Table
...
+      if headers.nil?
+        @table = array_of_rows
+        if @table.empty?
+          @headers = []
+        else
+          @headers = @table.first.headers
+        end
+      else
+        @headers = headers
+        @table = array_of_rows
```
Comment (esparta) [inline lib/csv/table.rb]: There's a duplication of code here: the `@table` assignment. We should keep only one at the beginning of `initialize`.

File: lib/csv/table.rb [hunk 5]
```diff
@@ -285,7 +301,8 @@ def delete_if(&block)
         @table.delete_if(&block)
       else                                      # by header
         deleted = []
-        headers.each do |header|
+        initial_headers = headers
```
Comment (esparta) [inline lib/csv/table.rb]: I can see you want to iterate through the initial headers of the table, but at this line `headers` is equals `@headers.dup`. Looks like we don't need this extra assignment.

File: lib/csv/table.rb [hunk 6]
```diff
@@ -23,8 +23,19 @@ class Table
     # * length()
     # * size()
     #
-    def initialize(array_of_rows)
-      @table = array_of_rows
+    def initialize(array_of_rows, headers=nil)
+      if headers.nil?
```
Comment (esparta) [inline lib/csv/table.rb]: This is only style, and also opinion: can we invert this to use ruby's truthy?
```ruby
if headers
  @headers = headers
else
  if @table.empty?
   # ...
  end
end
```

File: lib/csv/table.rb [hunk 7]
```diff
@@ -23,8 +23,19 @@ class Table
     # * length()
     # * size()
     #
-    def initialize(array_of_rows)
-      @table = array_of_rows
+    def initialize(array_of_rows, headers=nil)
```
Comment (esparta) [inline lib/csv/table.rb]: Since is a public class, can you add a documentation about the new parameter?

Comment (shibasaka) [reply lib/csv/table.rb]: I've added the documentation about headers argument.

File: lib/csv/table.rb (hunk 6 above)
Comment (shibasaka) [reply lib/csv/table.rb]: That it is clear, I've fixed it.

File: lib/csv/table.rb (hunk 5 above)
Comment (shibasaka) [reply lib/csv/table.rb]: You're right, I've fixed it.

File: lib/csv/table.rb (hunk 4 above)
Comment (shibasaka) [reply lib/csv/table.rb]: I've moved the @table assignment to the beginning of initialize.

**Comment by kou** [comment]
Thanks.
I've merged.

I've changed `headers` to keyword argument from optional argument and added a missing test to parse header only CSV.

**Comment by shibasaka** [comment]
Thanks!

**Pull request merged**